### PR TITLE
Updating and re-opening PR #48129

### LINKED ISSUE #43784: BUG: DatetimeIndex and Timestamp behave inconsistently with respect to pd.DateOffset
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
pd.DatetimeIndex([pd.to_datetime('2020-10-25').tz_localize('Europe/Berlin')]) + pd.DateOffset(hours=3) 
returns 
2020-10-25 03:00:00+01:00

while

pd.to_datetime('2020-10-25').tz_localize('Europe/Berlin') + pd.DateOffset(hours=3) 
returns 
Timestamp('2020-10-25 02:00:00+0100', tz='Europe/Berlin')
```

### Issue Description

Mismatch between the outputs

### Expected Behavior

Returning same datetime

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 73c68257545b5f8530b7044f56647bd2db92e2ba
python           : 3.9.4.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 165 Stepping 2, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252

pandas           : 1.3.3
numpy            : 1.19.5
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.2.4
setuptools       : 52.0.0.post20210125
Cython           : 0.29.23
pytest           : 6.2.3
hypothesis       : None
sphinx           : 4.0.1
blosc            : None
feather          : None
xlsxwriter       : 1.3.8
lxml.etree       : 4.6.3
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.3
IPython          : 7.22.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : 1.3.2
fsspec           : 0.9.0
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.4
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : None
pyxlsb           : 1.0.8
s3fs             : None
scipy            : 1.6.2
sqlalchemy       : 1.4.7
tables           : None
tabulate         : None
xarray           : None
xlrd             : 2.0.1
xlwt             : 1.3.0
numba            : 0.53.1

</details>

### PR DESCRIPTION
**Pull request by srotondo**
- [ ] closes #43784
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/arrays/datetimes.py [hunk 1]
```diff
@@ -761,7 +761,10 @@ def _add_offset(self, offset) -> DatetimeArray:
         assert not isinstance(offset, Tick)
 
         if self.tz is not None:
-            values = self.tz_localize(None)
+            if not offset._use_relativedelta:
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimes.py:764]: is there a viable way to handle this inside the Offset code?  in theory the DTA code shouldn't need to know anything about _use_relativedelta

File: pandas/tests/tseries/offsets/test_dst.py [hunk 2]
```diff
@@ -233,3 +235,30 @@ def test_nontick_offset_with_ambiguous_time_error(original_dt, target_dt, offset
...
+def test_series_dst_addition():
+    # GH#43784
+    startdates = pd.Series(
+        [
+            Timestamp("2020-10-25", tz="Europe/Berlin"),
+            Timestamp("2017-03-12", tz="US/Pacific"),
+        ]
+    )
+    offset1 = DateOffset(hours=3)
+    offset2 = DateOffset(days=1)
```
Comment (jbrockmendel) [inline pandas/tests/tseries/offsets/test_dst.py:249]: looks like this only tests the DateOffset class, for which _use_relativedelta can be True.  but the code you changed affects a bunch of other classes

File: pandas/core/arrays/datetimes.py (hunk 1 above)
Comment (srotondo) [reply pandas/core/arrays/datetimes.py:764]: The issue is with how these dates are localized, and I don't believe there is a good way to modify this in the Offset code.

File: pandas/tests/tseries/offsets/test_dst.py (hunk 2 above)
Comment (srotondo) [reply pandas/tests/tseries/offsets/test_dst.py:249]: Alright, I'll try and work on some tests for the other classes.

**Comment by srotondo** [comment]
Sorry this went stale, I've been really busy lately, I think I'll be able to find time to clean this up soon.

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**